### ISSUE TITLE
Add max-buffers=X and drop=True for appsink..?

### ISSUE THREAD

**Issue by normen** [fzwoch/obs-gstreamer#63]
I was wondering if it was possible to add the `max-buffer` and `drop` options for the appsink of gstreamer. I am getting a buildup of delay when OBS has issues draining the frames where I'd rather prefer to get dropouts.

Based on [this thread](https://forums.developer.nvidia.com/t/how-to-eliminate-gstreamer-camera-buffer/75608/6) it seems that gstreamer by default keeps frames that are not drained from the appsink in this buffer.

**Comment by fzwoch** [comment]
You can of course change the code and just add these options for your use case to the pipeline string and recompile. If you have the situation that appsink cannot push data into OBS fast enough (that should basically just be a GPU memory upload) without recovering from the jitter you basically stress the hardware too much. Perhaps it would make sense instead to change the pipeline to avoid the situation in the first place (e.g. `videorate`).

I haven't decide whether it is good to add these options or not. People just get too confused when dealing with too many options.

**Comment by normen** [comment]
I think its mainly a OBS issue, not really a system load issue. It seems OBS just "forgets" to drain the frames at some points, causing the buildup. It was more pronounced when obs-gstreamer was still blocking OBS when a pipeline was started, I'd basically get more delay with each started pipeline.

My issue is that while I manage to compile obs-gstreamer on Mac and Linux, my Windows build-fu is just too weak and my cameras are connected to a Windows machine. But I'll look into that more and try to build it myself.

**Comment by fzwoch** [comment]
The appsink really does not do much, it pushes data into OBS, it does not wait for OBS to "get" data.  may still be an issue with OBS when it cannot receive as much as you want to send - either way sounds like a bottleneck.

Not sure what the data look slike you want to push there and what your settings are - be aware that when you sync the appsinks to the clock they may also artificially cause the observed behavior.

**Comment by normen** [comment]
> The appsink really does not do much, it pushes data into OBS, it does not wait for OBS to "get" data. may still be an issue with OBS when it cannot receive as much as you want to send - either way sounds like a bottleneck.
> 
> Not sure what the data look slike you want to push there and what your settings are - be aware that when you sync the appsinks to the clock they may also artificially cause the observed behavior.

I don't sync, no. My whole gstreamer queue should have no buffers above 150ms but when "things happen" in OBS I get an increasing delay of up to a few seconds.. I can't pinpoint _when_ the delay increases but as said the old blocking version of this plugin caused it reliably.

This is my queue for getting one of the streams from a 4-channel HDMI encoder, the ntp stuff is there to make the inputs play in sync, then theres hardware decoding and a queue that basically just makes sure that there _is no_ queue.. The only place I can see the latency coming from is the appsink in this case..

`rtspsrc location="rtspt://192.168.2.19/main2" buffer-mode=0 ntp-sync=true ntp-time-source=0  latency=150 drop-on-latency=true  ! rtpmp2tdepay ! tsdemux ! h264parse ! nvh264dec ! watchdog timeout=2000 ! queue max-size-buffers=1 leaky=downstream ! video.`

**Comment by normen** [comment]
I finally managed to compile obs-gstreamer on windows. I added the max-buffers and drop options to the appsink as well as a fixed latency value for the pipeline and an NTP clock, that way I get perfectly synced cameras.

Idk if the buffers only play a role when theres overloads but I'd say that the latency and NTP clock options should be exposed through this plugin, it definitely makes sense to get good sync. Maybe I'll manage to create a PR at some point.
